builtin/logical/database/credentials.go

```diff
@@ -170,3 +175,187 @@ func (kg rsaKeyGenerator) configMap() (map[string]interface{}, error) {
 	}
 	return config, nil
 }
+
+type ClientCertificateGenerator struct {
+	CommonNameTemplate string `mapstructure:"common_name_template,omitempty"`
+	CAPrivateKey       string `mapstructure:"ca_private_key,omitempty"`
+	CACert             string `mapstructure:"ca_cert,omitempty"`
+	KeyType            string `mapstructure:"key_type,omitempty"`
+	KeyBits            int    `mapstructure:"key_bits,omitempty"`
+	SignatureBits      int    `mapstructure:"signature_bits,omitempty"`
+
+	parsedCABundle *certutil.ParsedCertBundle
+	cnProducer     template.StringTemplate
+}
+
+// newClientCertificateGenerator returns a new ClientCertificateGenerator
+// using the given config. Default values will be set on the returned
+// ClientCertificateGenerator if not provided in the config.
+func newClientCertificateGenerator(config map[string]interface{}) (ClientCertificateGenerator, error) {
+	var cg ClientCertificateGenerator
+	if err := mapstructure.WeakDecode(config, &cg); err != nil {
+		return cg, err
+	}
+
+	switch cg.KeyType {
+	case "":
+		cg.KeyType = "ec"
+	case "rsa":
+		switch cg.KeyBits {
+		case 0:
+			cg.KeyBits = 2048
+		case 2048, 3072, 4096:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ec":
+		switch cg.KeyBits {
+		case 0:
+			cg.KeyBits = 256
+		case 224, 256, 384, 521:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ed25519":
+	// key_bits ignored
+	default:
+		return cg, fmt.Errorf("invalid key_type")
+	}
+
+	switch cg.SignatureBits {
```

Comment:
We should set the default signature bits too when 0 is provided. Let's see if we can use the function recommended by Alex.